File: rs/pragmastat/src/fast_spread.rs
/// Fast O(n log n) implementation of the Spread (Shamos) estimator.
/// Based on Monahan's selection algorithm adapted for pairwise differences.
///
/// Internal implementation - not part of public API.
use crate::fnv1a::hash_f64_slice;
use crate::rng::Rng;

pub(crate) fn fast_spread(values: &[f64]) -> Result<f64, &'static str> {
    let n = values.len();
    if n <= 1 {
        return Ok(0.0);
    }
    if n == 2 {
        return Ok((values[1] - values[0]).abs());
    }

    // Validate for NaN/infinite values
    if values.iter().any(|v| !v.is_finite()) {
        return Err("Input contains NaN or infinite values");
    }

    if n > u32::MAX as usize {
        return Err("fast_spread: input too large");
    }

    // Sort the values
    let mut a = values.to_vec();
    a.sort_by(|x, y| x.total_cmp(y));

    // Total number of pairwise differences with i < j
    let total_pairs = (n as u64) * ((n - 1) as u64) / 2;
    let k_low = total_pairs.div_ceil(2); // 1-based rank of lower middle
    let k_high = (total_pairs + 2) / 2; // 1-based rank of upper middle

    // Per-row active bounds over columns j (0-based indices)
    // Row i allows j in [i+1, n-1] initially
    let mut left_bounds: Vec<u32> = (0..n).map(|i| (i + 1).min(n) as u32).collect();
    let mut right_bounds = vec![(n - 1) as u32; n];

    for i in 0..n {
        if left_bounds[i] > right_bounds[i] {
            left_bounds[i] = 1;
            right_bounds[i] = 0; // mark empty
        }
    }

    let mut row_counts: Vec<u32> = vec![0; n];

    // Initial pivot: a central gap
    let mut pivot = a[n / 2] - a[(n - 1) / 2];
    let mut prev_count_below = -1i64;

    let mut rng = Rng::from_seed(hash_f64_slice(values));

    loop {
        // === PARTITION: count how many differences are < pivot ===
        let mut count_below: u64 = 0;
        let mut largest_below = f64::NEG_INFINITY;
        let mut smallest_at_or_above = f64::INFINITY;

        let mut j = 1; // global two-pointer (non-decreasing across rows)
        for i in 0..n - 1 {
            if j < i + 1 {
                j = i + 1;
            }
            while j < n && a[j] - a[i] < pivot {
                j += 1;
            }

            let cnt_row = j.saturating_sub(i + 1);
            row_counts[i] = cnt_row as u32;
            count_below += cnt_row as u64;

            // boundary elements for this row
            if cnt_row > 0 {
                let cand_below = a[j - 1] - a[i];
                largest_below = largest_below.max(cand_below);
            }

            if j < n {
                let cand_at_or_above = a[j] - a[i];
                smallest_at_or_above = smallest_at_or_above.min(cand_at_or_above);
            }
        }

        // === TARGET CHECK ===
        let at_target = count_below == k_low || count_below == k_high - 1;

        if at_target {
            if k_low < k_high {
                // Even N: average the two central order stats
                return Ok(0.5 * (largest_below + smallest_at_or_above));
            } else {
                // Odd N: pick the single middle
                let need_largest = count_below == k_low;
                return Ok(if need_largest {
                    largest_below
                } else {
                    smallest_at_or_above
                });
            }
        }

        // === STALL HANDLING ===
        if count_below as i64 == prev_count_below {
            let mut min_active = f64::INFINITY;
            let mut max_active = f64::NEG_INFINITY;
            let mut active: u64 = 0;

            for i in 0..n - 1 {
                let li = left_bounds[i] as usize;
                let ri = right_bounds[i] as usize;
                if li > n || ri >= n {
                    return Err("fast_spread: bounds corrupted");
                }
                if li > ri {
                    continue;
                }

                let row_min = a[li] - a[i];
                let row_max = a[ri] - a[i];
                min_active = min_active.min(row_min);
                max_active = max_active.max(row_max);
                active += (ri - li + 1) as u64;
            }

            if active == 0 {
                if k_low < k_high {
                    return Ok(0.5 * (largest_below + smallest_at_or_above));
                }
                return Ok(if count_below >= k_low {
                    largest_below
                } else {
                    smallest_at_or_above
                });
            }

            if max_active <= min_active {
                return Ok(min_active);
            }

            let mid = 0.5 * (min_active + max_active);
            pivot = if mid > min_active && mid <= max_active {
                mid
            } else {
                max_active
            };
            prev_count_below = count_below as i64;
            continue;
        }

        // === SHRINK ACTIVE WINDOW ===
        if count_below < k_low {
            // Need larger differences: discard all strictly below pivot
            for i in 0..n - 1 {
                let i_u32 = i as u32;
                let new_left = i_u32
                    .checked_add(1)
                    .and_then(|v| v.checked_add(row_counts[i]))
                    .unwrap_or(n as u32); // overflow => mark empty below
                if new_left > left_bounds[i] {
                    left_bounds[i] = new_left;
                }
                if left_bounds[i] > right_bounds[i] {
                    left_bounds[i] = 1;
                    right_bounds[i] = 0;
                }
                let li = left_bounds[i] as usize;
                let ri = right_bounds[i] as usize;
                if li > n || ri >= n {
                    return Err("fast_spread: bounds corrupted");
                }
            }
        } else {
            // Too many below: keep only those strictly below pivot
            for i in 0..n - 1 {
                let i_u32 = i as u32;
                let new_right = match i_u32.checked_add(row_counts[i]) {
                    Some(v) => v,
                    None => {
                        left_bounds[i] = 1;
                        right_bounds[i] = 0;
                        continue;
                    }
                };
                if new_right < right_bounds[i] {
                    right_bounds[i] = new_right;
                }
                if right_bounds[i] < i_u32 + 1 {
                    left_bounds[i] = 1;
                    right_bounds[i] = 0;
                }
                let li = left_bounds[i] as usize;
                let ri = right_bounds[i] as usize;
                if li > n || ri >= n {
                    return Err("fast_spread: bounds corrupted");
                }
            }
        }

        prev_count_below = count_below as i64;

        // === CHOOSE NEXT PIVOT FROM ACTIVE SET ===
        let active_size: u64 = (0..n - 1)
            .filter(|&i| left_bounds[i] <= right_bounds[i])
            .map(|i| (right_bounds[i] - left_bounds[i] + 1) as u64)
            .sum();

        if active_size <= 2 {
            // Few candidates left: return midrange of remaining
            let mut min_rem = f64::INFINITY;
            let mut max_rem = f64::NEG_INFINITY;
            for i in 0..n - 1 {
                if left_bounds[i] > right_bounds[i] {
                    continue;
                }
                let lo = a[left_bounds[i] as usize] - a[i];
                let hi = a[right_bounds[i] as usize] - a[i];
                min_rem = min_rem.min(lo);
                max_rem = max_rem.max(hi);
            }

            if active_size == 0 {
                if k_low < k_high {
                    return Ok(0.5 * (largest_below + smallest_at_or_above));
                }
                return Ok(if count_below >= k_low {
                    largest_below
                } else {
                    smallest_at_or_above
                });
            }

            if k_low < k_high {
                return Ok(0.5 * (min_rem + max_rem));
            }
            // In this code path count_below < k_low, so min_rem is always the correct result:
            // |k_low-1-count_below| = d-1 <= d = |count_below-k_low| for all d > 0.
            return Ok(min_rem);
        } else {
            // Weighted random row selection
            let t = rng.uniform_i64(0, active_size as i64) as u64;
            let mut acc: u64 = 0;
            let mut row = 0;
            for r in 0..n - 1 {
                if left_bounds[r] > right_bounds[r] {
                    continue;
                }
                let size = (right_bounds[r] - left_bounds[r] + 1) as u64;
                if t < acc + size {
                    row = r;
                    break;
                }
                acc += size;
            }

            // Median column of the selected row
            let left = left_bounds[row] as usize;
            let right = right_bounds[row] as usize;
            if left >= n || right >= n || left > right {
                return Err("fast_spread: active bounds out of range");
            }
            let col = (left + right) / 2;
            if col >= n {
                return Err("fast_spread: pivot index out of range");
            }
            pivot = a[col] - a[row];
        }
    }
}

File: kt/src/main/kotlin/dev/pragmastat/FastSpread.kt
package dev.pragmastat

import kotlin.math.abs
import kotlin.random.Random

/**
 * Fast O(n log n) implementation of the Spread (Shamos) estimator.
 * Based on Monahan's selection algorithm adapted for pairwise differences.
 *
 * Internal implementation - not part of public API.
 */
internal fun fastSpread(values: List<Double>): Double {
    val n = values.size
    require(n > 0) { "Input list cannot be empty" }
    if (n == 1) return 0.0
    if (n == 2) return abs(values[1] - values[0])

    val a = values.sorted()
    val N = (n.toLong() * (n - 1)) / 2
    val kLow = (N + 1) / 2
    val kHigh = (N + 2) / 2

    val L = IntArray(n) { i -> minOf(i + 1, n) }
    val R = IntArray(n) { n - 1 }

    for (i in 0 until n) {
        if (L[i] > R[i]) {
            L[i] = 1
            R[i] = 0
        }
    }

    val rowCounts = IntArray(n)
    var pivot = a[n / 2] - a[(n - 1) / 2]
    var prevCountBelow = -1L

    while (true) {
        var countBelow = 0L
        var largestBelow = Double.NEGATIVE_INFINITY
        var smallestAtOrAbove = Double.POSITIVE_INFINITY

        var j = 1
        for (i in 0 until n - 1) {
            if (j < i + 1) j = i + 1
            while (j < n && a[j] - a[i] < pivot) j++

            val cntRow = maxOf(0, j - (i + 1))
            rowCounts[i] = cntRow
            countBelow += cntRow

            if (cntRow > 0) largestBelow = maxOf(largestBelow, a[j - 1] - a[i])
            if (j < n) smallestAtOrAbove = minOf(smallestAtOrAbove, a[j] - a[i])
        }

        val atTarget = countBelow == kLow || countBelow == kHigh - 1

        if (atTarget) {
            return if (kLow < kHigh) {
                0.5 * (largestBelow + smallestAtOrAbove)
            } else {
                if (countBelow == kLow) largestBelow else smallestAtOrAbove
            }
        }

        if (countBelow == prevCountBelow) {
            var minActive = Double.POSITIVE_INFINITY
            var maxActive = Double.NEGATIVE_INFINITY
            var active = 0

            for (i in 0 until n - 1) {
                if (L[i] > R[i]) continue
                minActive = minOf(minActive, a[L[i]] - a[i])
                maxActive = maxOf(maxActive, a[R[i]] - a[i])
                active += R[i] - L[i] + 1
            }

            if (active <= 0) {
                return if (kLow < kHigh) {
                    0.5 * (largestBelow + smallestAtOrAbove)
                } else {
                    if (countBelow >= kLow) largestBelow else smallestAtOrAbove
                }
            }

            if (maxActive <= minActive) return minActive

            val mid = 0.5 * (minActive + maxActive)
            pivot = if (mid > minActive && mid <= maxActive) mid else maxActive
            prevCountBelow = countBelow
            continue
        }

        if (countBelow < kLow) {
            for (i in 0 until n - 1) {
                val newL = i + 1 + rowCounts[i]
                if (newL > L[i]) L[i] = newL
                if (L[i] > R[i]) {
                    L[i] = 1
                    R[i] = 0
                }
            }
        } else {
            for (i in 0 until n - 1) {
                val newR = i + rowCounts[i]
                if (newR < R[i]) R[i] = newR
                if (R[i] < i + 1) {
                    L[i] = 1
                    R[i] = 0
                }
            }
        }

        prevCountBelow = countBelow

        val activeSize = (0 until n - 1).filter { L[it] <= R[it] }.sumOf { (R[it] - L[it] + 1).toLong() }

        if (activeSize <= 2) {
            var minRem = Double.POSITIVE_INFINITY
            var maxRem = Double.NEGATIVE_INFINITY
            for (i in 0 until n - 1) {
                if (L[i] > R[i]) continue
                minRem = minOf(minRem, a[L[i]] - a[i])
                maxRem = maxOf(maxRem, a[R[i]] - a[i])
            }

            if (activeSize <= 0) {
                return if (kLow < kHigh) {
                    0.5 * (largestBelow + smallestAtOrAbove)
                } else {
                    if (countBelow >= kLow) largestBelow else smallestAtOrAbove
                }
            }

            return if (kLow < kHigh) {
                0.5 * (minRem + maxRem)
            } else {
                if (abs((kLow - 1) - countBelow) <= abs(countBelow - kLow)) minRem else maxRem
            }
        } else {
            val t = Random.nextLong(activeSize)
            var acc = 0L
            var row = 0
            for (r in 0 until n - 1) {
                if (L[r] > R[r]) continue
                val size = R[r] - L[r] + 1
                if (t < acc + size) {
                    row = r
                    break
                }
                acc += size
            }

            val col = (L[row] + R[row]) / 2
            pivot = a[col] - a[row]
        }
    }
}
